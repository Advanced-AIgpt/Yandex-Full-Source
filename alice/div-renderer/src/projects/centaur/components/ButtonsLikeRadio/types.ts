import { Div, IDivAction } from 'divcard2';
import { IRequestState } from '../../../../common/types/common';
import { IStatePlace } from '../../../../common/actions/div';
import { DivStateBlockProps } from '../../helpers/types';

type createDivWithAction = (actions?: IDivAction[]) => Div;

interface IButtonLikeRadioProps {
    activeDiv: createDivWithAction;
    inactiveDiv: createDivWithAction;
}

export interface IButtonsLikeRadioProps {
    buttons: IButtonLikeRadioProps[];
    requestState: IRequestState;
    buttonListId: string;
    startActive?: number;
    places?: IStatePlace[];
    options?: Omit<DivStateBlockProps, 'states' | 'default_state_id' | 'div_id'>;
}
