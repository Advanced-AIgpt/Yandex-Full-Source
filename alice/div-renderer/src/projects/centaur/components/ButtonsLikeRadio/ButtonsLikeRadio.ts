import { ContainerBlock, Div, DivStateBlock } from 'divcard2';
import { setStateActionInAllPlaces, setVariable } from '../../../../common/actions/div';
import { IButtonsLikeRadioProps } from './types';

export function getButtonListVariableId(buttonListId: string) {
    return `buttonList_${buttonListId}`;
}

export default function ButtonsLikeRadio({
    buttons,
    buttonListId,
    startActive = 0,
    requestState,
    places = [{ name: 'top_level', place: ['0'] }],
    options,
}: IButtonsLikeRadioProps): Div[] {
    requestState.variables.add({
        type: 'number',
        name: getButtonListVariableId(buttonListId),
        value: startActive,
    });

    return buttons.map((el, index) => {
        const stateId = `${buttonListId}_button_${index}`;
        const activeId = `${buttonListId}_button_${index}_active`;
        const inactiveId = `${buttonListId}_button_${index}_inactive`;

        requestState.variableTriggers.add({
            condition: `@{${getButtonListVariableId(buttonListId)} == ${index}.0}`,
            actions: setStateActionInAllPlaces({
                places,
                logPrefix: `Set button ${index} of button list ${buttonListId} to active`,
                state: [
                    stateId,
                    activeId,
                ],
            }),
        });
        requestState.variableTriggers.add({
            condition: `@{${getButtonListVariableId(buttonListId)} != ${index}.0}`,
            actions: setStateActionInAllPlaces({
                places,
                logPrefix: `Set button ${index} of button list ${buttonListId} to inactive`,
                state: [
                    stateId,
                    inactiveId,
                ],
            }),
        });

        return new ContainerBlock({
            items: [
                new DivStateBlock({
                    ...options,
                    div_id: stateId,
                    default_state_id: startActive === index ? activeId : inactiveId,
                    states: [
                        {
                            state_id: activeId,
                            div: el.activeDiv([]),
                        },
                        {
                            state_id: inactiveId,
                            div: el.inactiveDiv([
                                {
                                    log_id: 'set active variable',
                                    url: setVariable(getButtonListVariableId(buttonListId), index),
                                },
                            ]),
                        },
                    ],
                }),
            ],
        });
    });
}
