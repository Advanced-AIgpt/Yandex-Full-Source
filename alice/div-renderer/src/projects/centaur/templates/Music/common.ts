import {
    ContainerBlock,
    Div,
    FixedSize,
    IDivAction,
    ImageBackground,
    ImageBlock, MatchParentSize,
    SolidBackground,
    TextBlock,
} from 'divcard2';
import { colorBlackOpacity70, offsetFromEdgeOfScreen } from '../../style/constants';
import { compact } from 'lodash';
import EmptyDiv from '../../components/EmptyDiv';
import { text28m } from '../../style/Text/Text';
import { directivesAction } from '../../../../common/actions';
import { centaurMusicPlay } from '../../../../common/actions/server/musicActions';
import MusicPlaying from './MusicPlaying/MusicPlaying';
import { IRequestState } from '../../../../common/types/common';
import LikeButton from './components/LikeButton';

interface IGetPlaylistItemProps {
    title: string;
    subtitle?: string;
    objectId: string;
    objectType: string;
    startFromTrackId?: string;
    isCurrent?: boolean;
    imageUrl: string;
    imageBackground?: string;
    requestState: IRequestState;
    isLiked?: boolean;
}

function CoverWithBackground(imageUrl: string, background: string, action: IDivAction, isCurrent: boolean, requestState: IRequestState) {
    return new ContainerBlock({
        width: new FixedSize({ value: 80 }),
        height: new FixedSize({ value: 80 }),
        alignment_vertical: 'top',
        actions: [action],
        margins: {
            right: 24,
        },
        border: {
            corner_radius: 40,
        },
        content_alignment_horizontal: 'center',
        content_alignment_vertical: 'center',
        background: compact([
            new SolidBackground({
                color: background,
            }),
        ]),
        orientation: 'overlap',
        items: compact([
            new ImageBlock({
                image_url: imageUrl,
                width: new FixedSize({ value: 48 }),
                height: new FixedSize({ value: 48 }),
            }),
            isCurrent && new EmptyDiv({
                width: new FixedSize({ value: 80 }),
                height: new FixedSize({ value: 80 }),
                background: [
                    new SolidBackground({
                        color: colorBlackOpacity70,
                    }),
                ],
            }),
            isCurrent && MusicPlaying(requestState),
        ]),
    });
}

function CoverWithoutBackground(imageUrl: string, action: IDivAction, isCurrent: boolean, requestState: IRequestState) {
    return new ContainerBlock({
        width: new FixedSize({ value: 80 }),
        height: new FixedSize({ value: 80 }),
        alignment_vertical: 'top',
        content_alignment_horizontal: 'center',
        content_alignment_vertical: 'center',
        actions: [action],
        margins: {
            right: 24,
        },
        border: {
            corner_radius: 20,
        },
        background: compact([
            new ImageBackground({
                image_url: imageUrl,
            }),
            isCurrent && new SolidBackground({
                color: colorBlackOpacity70,
            }),
        ]),
        items: [
            isCurrent ? MusicPlaying(requestState) : new EmptyDiv(),
        ],
    });
}

export function getPlaylistItem({
    isCurrent = false,
    objectType,
    subtitle,
    objectId,
    title,
    startFromTrackId,
    imageUrl,
    imageBackground,
    requestState,
    isLiked,
}: IGetPlaylistItemProps): {
    options: Partial<ConstructorParameters<typeof ContainerBlock>[0]>;
    items: Div[];
} {
    const openAction = {
        log_id: 'play_music_click',
        url: directivesAction(centaurMusicPlay({
            disableNlg: true,
            objectType,
            objectId,
            startFromTrackId,
        })),
    };

    return {
        options: {
            orientation: 'horizontal',
            paddings: {
                top: 28,
                left: offsetFromEdgeOfScreen,
                bottom: 28,
                right: offsetFromEdgeOfScreen,
            },
            content_alignment_vertical: 'center',
        },
        items: compact([
            imageBackground ?
                CoverWithBackground(imageUrl, imageBackground, openAction, isCurrent, requestState) :
                CoverWithoutBackground(imageUrl, openAction, isCurrent, requestState),
            new ContainerBlock({
                width: new MatchParentSize({ weight: 1 }),
                actions: [openAction],
                items: compact([
                    new TextBlock({
                        ...text28m,
                        text: title,
                    }),
                    subtitle && new TextBlock({
                        ...text28m,
                        text: subtitle,
                        alpha: 0.5,
                    }),
                ]),
            }),
            typeof isLiked !== 'undefined' && LikeButton({ isLiked, trackId: startFromTrackId || '' }),
        ]),
    };
}
