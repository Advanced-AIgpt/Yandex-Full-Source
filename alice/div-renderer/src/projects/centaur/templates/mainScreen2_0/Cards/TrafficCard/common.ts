import { ITrafficForecast } from './types';

export function sortTimeOfDay(originalArray: Readonly<ITrafficForecast[]>): ITrafficForecast[] {
    const forecastsFirst = [];
    const forecastsLast = [];

    for (const forecastElement of originalArray) {
        if (forecastElement.time === null || typeof forecastElement.time === 'undefined') {
            continue;
        }

        if (forecastsFirst.length === 0) {
            forecastsFirst.push(forecastElement);
        } else if (forecastsLast.length > 0) {
            forecastsLast.push(forecastElement);
        } else if (forecastElement.time - (forecastsFirst[forecastsFirst.length - 1].time as number) === 1) {
            forecastsFirst.push(forecastElement);
        } else {
            forecastsLast.push(forecastElement);
        }
    }

    return [...forecastsLast, ...forecastsFirst];
}
