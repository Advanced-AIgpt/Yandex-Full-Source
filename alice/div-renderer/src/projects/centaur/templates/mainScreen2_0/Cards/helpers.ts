export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
