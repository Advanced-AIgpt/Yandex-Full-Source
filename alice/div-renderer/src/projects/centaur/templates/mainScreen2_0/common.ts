import { IDivAction } from 'divcard2';
import { EnumLayer } from '../../actions/client';
import { CrutchLoader } from '../../components/Loader/Loader';

interface IPutLoaderToActionOptions {
    action?: string | null,
    layer: EnumLayer,
    actionLogId?: string,
}

export const putLoaderToAction = ({
    action,
    layer,
    actionLogId = 'action_with_loader',
}: IPutLoaderToActionOptions): IDivAction[] | undefined => {
    if (!action) {
        return undefined;
    }
    return [
        CrutchLoader(layer),
        {
            log_id: actionLogId,
            url: action,
        },
    ];
};
