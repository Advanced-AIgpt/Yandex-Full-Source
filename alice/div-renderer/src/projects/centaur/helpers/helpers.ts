/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
