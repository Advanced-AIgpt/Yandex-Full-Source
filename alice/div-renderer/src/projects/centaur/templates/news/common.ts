import {
    ContainerBlock,
    DivBackground,
    FixedSize, IDivAction,
    ImageBackground,
    ImageBlock,
    MatchParentSize,
    SolidBackground,
    TextBlock,
    WrapContentSize,
} from 'divcard2';
import { compact } from 'lodash';
import { utcToZonedTime } from 'date-fns-tz';
import { differenceInCalendarDays, isSameDay } from 'date-fns';
import { Long } from 'protobufjs';
import { NAlice } from '../../../../protos';
import { formatDate } from '../../helpers/helpers';
import {
    colorDarkGrey1Opacity70,
    colorReadButton,
    colorWhite,
    colorWhiteOpacity60,
    offsetFromEdgeOfScreen,
} from '../../style/constants';
import { text40r, title32m, title60m } from '../../style/Text/Text';
import { getImageByTopic } from '../../helpers/newsHelper';
import { Avatar } from '../../../../common/helpers/avatar';

export const patchToOrig = (avatarsUrl: string) => avatarsUrl.replace(/\/[^/]+$/, '/orig');

export type NewsCardProps = {
    item: NAlice.NData.ITNewsItem;
    topic: string;
    bottomGap: boolean;
    tz: string;
};

export function newsCommonCard(
    { item: { Text, Image, Agency, Logo, PubDate, ExtendedNews }, topic, bottomGap, tz }: NewsCardProps,
    openExtendedNewsDivAction: IDivAction) {
    return new ContainerBlock({
        width: new MatchParentSize(),
        height: new MatchParentSize(),
        background: generateBackground(topic, Image),
        content_alignment_horizontal: 'left',
        content_alignment_vertical: 'bottom',
        paddings: {
            left: offsetFromEdgeOfScreen,
            bottom: bottomGap ? 104 : offsetFromEdgeOfScreen,
            right: offsetFromEdgeOfScreen,
        },
        items: [
            new ContainerBlock({
                width: new WrapContentSize(),
                height: new MatchParentSize({ weight: 1 }),
                content_alignment_vertical: 'bottom',
                orientation: 'horizontal',
                margins: {
                    bottom: 20,
                },
                items: newsSource(Agency, Logo, PubDate, tz),
            }),
            new ContainerBlock({
                width: new MatchParentSize(),
                height: new WrapContentSize(),
                content_alignment_vertical: 'bottom',
                orientation: 'horizontal',
                items: compact([
                    new TextBlock({
                        ...title60m,
                        width: new MatchParentSize({ weight: 1 }),
                        height: new WrapContentSize(),
                        max_lines: 5,
                        truncate: 'end',
                        text: Text ?? ' ',
                    }),
                    ExtendedNews?.length && new TextBlock({
                        ...text40r,
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                        background: [new SolidBackground({
                            color: colorReadButton,
                        })],
                        border: {
                            corner_radius: 24,
                        },
                        paddings: {
                            top: 22,
                            right: 33,
                            bottom: 22,
                            left: 33,
                        },
                        margins: {
                            left: 48,
                        },
                        alignment_horizontal: 'right',
                        text: 'Читать',
                        action: openExtendedNewsDivAction,
                    }),
                ]),
            }),
        ],
    });
}

export const newsSource = (
    Agency: (string | null | undefined),
    Logo: (string | null | undefined),
    PubDate: (number | Long | null | undefined),
    tz: string,
) =>
    compact([
        Logo && new ContainerBlock({
            background: [
                new SolidBackground({
                    color: colorWhite,
                }),
            ],
            content_alignment_horizontal: 'center',
            content_alignment_vertical: 'center',
            border: {
                corner_radius: 12,
            },
            width: new FixedSize({ value: 40 }),
            height: new FixedSize({ value: 40 }),
            margins: {
                right: 20,
            },
            items: [
                new ImageBlock({
                    width: new FixedSize({ value: 28 }),
                    height: new FixedSize({ value: 28 }),
                    image_url: Logo ?? '',
                }),
            ],
        }),
        new TextBlock({
            ...title32m,
            text_color: colorWhiteOpacity60,
            width: new WrapContentSize(),
            height: new WrapContentSize(),
            text_alignment_vertical: 'center',
            text: (Agency ? Agency + ' • ' : '') + (Number(PubDate) ? formDate(Number(PubDate) * 1000 || 0, tz) : ''),
        }),
    ]);

export const generateBackground = (
    Topic: string | null | undefined,
    NewsImage: NAlice.NData.TNewsItem.ITImage | null | undefined,
    color = colorDarkGrey1Opacity70,
) => {
    const result: DivBackground[] = [];
    if (NewsImage && NewsImage.Src) {
        const avatar = Avatar.fromUrl(NewsImage.Src);

        result.push(
            new ImageBackground({
                image_url: avatar ? avatar.setTypeName('orig').toString() : NewsImage.Src,
                scale: 'fill',
                preload_required: 1,
            }),
        );
    } else {
        result.push(
            new ImageBackground({
                image_url: getImageByTopic(Topic),
                preload_required: 1,
            }),
        );
    }

    result.push(new SolidBackground({ color }));
    return result;
};

export const formDate = (date: number, tz: string) => {
    const d = utcToZonedTime(date, tz);
    const now = utcToZonedTime(Date.now(), tz);
    if (isSameDay(d, now)) {
        return formatDate(d, 'HH:mm');
    }
    if (differenceInCalendarDays(now, d) === 1) {
        return `вчера в ${formatDate(d, 'HH:mm')}`;
    }
    return formatDate(d, 'd MMMM в HH:mm');
};
