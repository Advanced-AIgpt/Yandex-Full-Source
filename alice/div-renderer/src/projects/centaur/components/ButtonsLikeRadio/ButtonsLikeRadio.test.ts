import ButtonsLikeRadio from './ButtonsLikeRadio';
import { TextBlock } from 'divcard2';
import { MMRequest } from '../../../../common/helpers/MMRequest';
import { createRequestState } from '../../../../registries/common';
import { AnonymizeDataForSnapshot } from '../../../../common/helpers/dev';

describe('Check ButtonsLikeRadio render', () => {
    it('should success render', () => {
        const mmRequest = new MMRequest({}, {}, {});
        const requestState = createRequestState(mmRequest);

        expect(AnonymizeDataForSnapshot(ButtonsLikeRadio({
            requestState,
            buttonListId: 'id',
            buttons: [
                {
                    activeDiv: actions => new TextBlock({
                        text: 'active',
                        actions,
                    }),
                    inactiveDiv: actions => new TextBlock({
                        text: 'inactive',
                        actions,
                    }),
                },
            ],
        }))).toMatchSnapshot();
    });
});
