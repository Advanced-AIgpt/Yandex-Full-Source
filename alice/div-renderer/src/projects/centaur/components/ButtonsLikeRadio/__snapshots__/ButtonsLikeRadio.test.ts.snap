// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check ButtonsLikeRadio render should success render 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "default_state_id": "id_button_0_active",
        "div_id": "id_button_0",
        "states": Array [
          Object {
            "div": Object {
              "actions": Array [],
              "text": "active",
              "type": "text",
            },
            "state_id": "id_button_0_active",
          },
          Object {
            "div": Object {
              "actions": Array [
                Object {
                  "log_id": "set active variable",
                  "url": "div-action://set_variable?name=buttonList_id&value=0",
                },
              ],
              "text": "inactive",
              "type": "text",
            },
            "state_id": "id_button_0_inactive",
          },
        ],
        "type": "state",
      },
    ],
    "type": "container",
  },
]
`;
